### ISSUE TITLE
stdlib/STL_stdlib.c causes MSVC error C2268 when LTO is enabled

### ISSUE THREAD

**Issue by jonesmz** [libsdl-org/SDL#5732]
Version 2.0.22, as well as with commit https://github.com/libsdl-org/SDL/commit/0e63b514d505116766c96a87ac8c5d614fe37cf1

The problem is described here: https://docs.microsoft.com/en-us/cpp/error-messages/compiler-errors-1/compiler-error-c2268?view=msvc-170

The issue can be reproduced by having a project which includes SDL2 as a git-submodule and `include()`'s SDL2's CMakeLists.txt in it's build tree.

When you have

```
CMAKE_POLICY(SET CMP0069 NEW)
set(CMAKE_POLICY_DEFAULT_CMP0069 NEW)

includeCheckIPOSupported)
check_ipo_supported(RESULT lto_supported OUTPUT error)
if( lto_supported )
message(STATUS "IPO / LTO enabled")
set(CMAKE_INTERPROCEDURAL_OPTIMIZATION TRUE)
else()
    message(STATUS "IPO / LTO not supported: <${error}>")
endif()

include(SDL2)
.....

build your project here
```

You get errors like:
> D:\a\osp-magnum\osp-magnum\3rdparty\SDL2\src\stdlib\SDL_string.c(301): error C2268: 'memset' is a compiler predefined library helper. Library helpers are not supported with /GL; compile object file 'd:\a\osp-magnum\osp-magnum\build\3rdparty\sdl2\sdl2.dir\release\sdl_stdlib.obj' without /GL. [D:\a\osp-magnum\osp-magnum\build\3rdparty\SDL2\SDL2.vcxproj]

I added the following to SDL2\CMakeLists.txt around line 490 (as of release 2.0.22, anyway), and it made the problem go away.

```
if(MSVC)
    set_source_files_properties(${SDL2_SOURCE_DIR}/src/stdlib/SDL_stdlib.c PROPERTIES COMPILE_FLAGS /GL-)
endif()
```

A more elegant solution might use generator expressions to determine if the file is being compiled with LTO enabled or not, before passing the /GL- flag to it.

**Comment by madebr** [comment]
You mean something as follows?
```cmake
if(MSVC)
    set_source_files_properties(${SDL2_SOURCE_DIR}/src/stdlib/SDL_stdlib.c PROPERTIES COMPILE_FLAGS "$<$<TARGET_PROPERTY:INTERPROCEDURAL_OPTIMIZATION>:/GL->")
endif()
```

**Comment by jonesmz** [comment]
Yes, something like that generator expression.

But I don't think you can do `$<TARGET_PROPERTY:INTERPROCEDURAL_OPTIMIZATION>` without explicitly specifying a target to evaluate, can you? Does cmake let you omit the target to evaluate, and if you omit it, it evaluates based on the target that the source file is being built for?

The difficulty that I had with expressing a full solution was that the file is built for both the SDL target (dynamic library), as well as the SDL_static target (static library), and I didn't want to make the assumption that the dynamic and static library versions would always be built with the same link time optimization settings.

I'm also not certain that specifically singling out this one source file to not be built with LTO is the right approach. My main issue with being able to give a good recommendation on the right solution is that this behavior from the MSVC compiler is... bizarre. It almost seems to me that the non-hacky solution is to disable the generation of these special "library helper" functions in the SDL library entirely, in favor of using the built-in intrinsic instead, when LTO is enabled.

Another thing to be careful of is that even though this specific source file is apparently the only one that causes a problem, you also have to explicitly remove it from UNITY builds to avoid the /GL- flag from being dropped, or possibly infecting the other source files that are part of the unity build.

This doesn't seem to be something that the SDL cmake files takes into consideration, but Unity builds are only available in the very newest cmake releases, so might be that not many systems have it available.

**Comment by madebr** [comment]
When you do `set(CMAKE_INTERPROCEDURAL_OPTIMIZATION TRUE)`, the `INTERPROCEDURAL_OPTIMIZATION` property will be enabled on all targets created after that `set`.
Which means both `SDL2` and `SDL2-static` have it enabled.

So we can simply do:
```cmake
if(MSVC)
  if(CMAKE_INTERPROCEDURAL_OPTIMIZATION)
     set_property(SOURCE src/stdlib/SDL_stdlib.c APPEND PROPERTY COMPILE_FLAGS /GL-)
   endif()
endif()
```
We don't need to check for IPO support because `CMAKE_INTERPROCEDURAL_OPTIMAZATION` was enabled.

The compiler emits the following warning, which I don't know how to disable:
```
cl : command line warning D9025: overriding '/GL' with '/GL-'
```

[This answer](https://developercommunity.visualstudio.com/t/cant-suppress-compiler-intrinsic-function-calls-wh/865266#T-N884309) gives a tip-of-the-iceberg explanation about why we're seeing this error.

Since `SDL_stdlib.c` contains a lot of little functions which are very useful to have LTO enabled for which the linker can then practically inline,
I propose to move [this block](https://github.com/libsdl-org/SDL/blob/4ddf85b1d5e6930766d681fd298220109ffda311/src/stdlib/SDL_stdlib.c#L554-L574) to a dedicated source file and compile that file with LTO disabled.

About unity builds, I just tried to build with `-DCMAKE_UNITY_BUILD=ON` passed to CMake, and I got a big huge error log.
So we might need to fix that first (I don't have much experience with them).
CMake has [`SKIP_UNITY_BUILD_INCLUSION`](https://cmake.org/cmake/help/latest/prop_sf/SKIP_UNITY_BUILD_INCLUSION.html) though.
So when we extract the intrinsic functions to a separate file, we can exclude them from the unity build.

**Comment by jonesmz** [comment]
top level cmakelists.txt can change the properties of targets after they've been initialized by the CMakeLists.txt that sets them up.

That's why I said it wasn't necessarily the case that we can assume SDL is IPO, but SDL-static is also IPO.

But if you want to consider it an error to compile SDL and SDL-static with different IPO levels, that's fine with me. I personally don't see why I would ever want that, though I can imagine someone wanting to for some reason or another.

**Comment by madebr** [comment]
I think there is value in providing support for building SDL2 with LTO,
as it should result in a faster/more performant library.
The file size of a `SDL2.dll`  grows from 1.37 MB to 1.44 MB (+70kB) when enabling LTO.
I can't tell anything about performance though.


About a unity build, adding support for this is not so trivial.
Enabling `CMAKE_UNITY_BUILD` gives a lot of errors.
This is because SDL provides support for a lot of subsystems that use similar names for private types and static variables
 SDL is also C-only, so we cannot [use namespaces to work around this](https://cmake.org/cmake/help/latest/prop_tgt/UNITY_BUILD_UNIQUE_ID.html#unity-build-unique-id).
We could of course group the build by using [`UNITY_GROUP`](https://cmake.org/cmake/help/latest/prop_sf/UNITY_GROUP.html), but imho the impact on the cmake script might be too big.

**Comment by jonesmz** [comment]
Issue appears fixed in prerelease-2.23.1 for my use case.

https://github.com/TheOpenSpaceProgram/osp-magnum/pull/212

Thank you very much.
